src/linuxthreads.cc

```diff
@@ -226,49 +226,63 @@ struct ListerParams {
   char *altstack_mem;
   ListAllProcessThreadsCallBack callback;
   void *parameter;
+  pid_t pid;
   va_list ap;
 };
 
 static void ListerThread(struct ListerParams *args) {
   int found_parent = 0;
   pid_t clone_pid = sys_gettid(), ppid = sys_getppid();
-  char proc_self_task[80], marker_name[48], *marker_path;
+  char proc_task[80], marker_name[48], *marker_path;
   const char *proc_paths[3];
   const char *const *proc_path = proc_paths;
   int proc = -1, marker = -1, num_threads = 0;
   int max_threads = 0, sig;
   struct kernel_stat marker_sb, proc_sb;
   stack_t altstack;
+  pid_t expid = args->pid;
 
+  marker_path = NULL;
   /* Create "marker" that we can use to detect threads sharing the same
    * address space and the same file handles. By setting the FD_CLOEXEC flag
    * we minimize the risk of misidentifying child processes as threads;
    * and since there is still a race condition,  we will filter those out
    * later, anyway.
    */
-  if ((marker = sys_socket(PF_LOCAL, SOCK_DGRAM, 0)) < 0 || sys_fcntl(marker, F_SETFD, FD_CLOEXEC) < 0) {
-  failure:
-    args->result = -1;
-    args->err = errno;
-    if (marker >= 0) NO_INTR(sys_close(marker));
-    sig_marker = marker = -1;
-    if (proc >= 0) NO_INTR(sys_close(proc));
-    sig_proc = proc = -1;
-    sys__exit(1);
-  }
+  if (expid == 0) {
+    if ((marker = sys_socket(PF_LOCAL, SOCK_DGRAM, 0)) < 0 || sys_fcntl(marker, F_SETFD, FD_CLOEXEC) < 0) {
+      goto failure;
+    }
 
   /* Compute search paths for finding thread directories in /proc            */
-  local_itoa(strrchr(strcpy(proc_self_task, "/proc/"), '\000'), ppid);
-  strcpy(marker_name, proc_self_task);
-  marker_path = marker_name + strlen(marker_name);
-  strcat(proc_self_task, "/task/");
-  proc_paths[0] = proc_self_task; /* /proc/$$/task/                          */
-  proc_paths[1] = "/proc/";       /* /proc/                                  */
-  proc_paths[2] = NULL;
-
+    local_itoa(strrchr(strcpy(proc_task, "/proc/"), '\000'), ppid);
+    strcpy(marker_name, proc_task);
+    marker_path = marker_name + strlen(marker_name);
   /* Compute path for marker socket in /proc                                 */
-  local_itoa(strcpy(marker_path, "/fd/") + 4, marker);
-  if (sys_stat(marker_name, &marker_sb) < 0) {
+    local_itoa(strcpy(marker_path, "/fd/") + 4, marker);
+    if (sys_stat(marker_name, &marker_sb) < 0) {
+      goto failure;
+    }
+    strcat(proc_task, "/task/");
+    proc_paths[0] = proc_task; /* /proc/$$/task/                          */
+    proc_paths[1] = "/proc/";       /* /proc/                                  */
+    proc_paths[2] = NULL;
+  } else {
+    /* Markers cannot be used to differentiate between
+     * process and thread, if we are tracing external process.
+     * Hence we cannot iterate /proc/ to find thread belong our
```

Comment:
I don't think we need to worry about this do we? We're not going to be on a kernel w/o /proc/<>/task

Reply:
No, we are not going to iterate the /proc directory, hence we are not adding it to proc_paths. Thats why added this comment.